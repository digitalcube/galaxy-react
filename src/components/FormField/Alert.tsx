import React, { FC, PropsWithChildren } from "react";
import { FaExclamationCircle } from "react-icons/fa";


export type FormAlertProps = PropsWithChildren<{
    errorMessage?: string;
}>
export const FormAlert: FC<FormAlertProps> = ({errorMessage, children}) => {
    if (!errorMessage && !children) return null;
    return (
        <div className="alert alert-danger mb-4 p-0 border-0 d-flex c-alert" role="alert">
            <div className="alert-icon px-1 d-flex align-items-center">
                <i aria-hidden="true">
                    <FaExclamationCircle className="d-block" />
                </i>
            </div>
            <div className="alert-body px-3 py-2 flex-grow-1">
                <p className="mb-0">{errorMessage}</p>
                {children}
            </div>
        </div>
    )
}