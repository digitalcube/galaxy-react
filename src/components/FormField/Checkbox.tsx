import React , {FC} from 'react'
import { FormGroup, Input, InputProps, Label } from 'reactstrap'


export type CheckboxProps = {
    label: string;
    id: string;
} & Omit<InputProps, 'type'>
export const Checkbox: FC<CheckboxProps> = (props) => {
    return (
        <FormGroup className="form-check d-flex align-items-center colored-check smaller">
            <Input {...props} type="checkbox" id={props.id} className="screen-reader-text" checked={props.value === true} />
            <Label className="form-check-label" htmlFor={props.id}>{props.label}</Label>
        </FormGroup>
    )
}