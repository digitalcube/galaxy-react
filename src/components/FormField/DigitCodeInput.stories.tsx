import React from 'react';
import { Meta, Story } from '@storybook/react';
import {
  DigitCodeInput,
  DigitCodeInputProps,
  UncontrolledDigitCodeInput,
  UncontrolledDigitCodeInputProps,
} from './DigitCodeInput';

const meta: Meta = {
  title: 'Components/Forms/DigitCodeInput',
  component: DigitCodeInput,
  parameters: {
    controls: { expanded: true },
  },
};

export default meta;

const Template: Story<DigitCodeInputProps> = (args) => (
  <>
    <DigitCodeInput {...args} />
  </>
);

// By passing using the Args format for exported stories, you can control the props for a component for reuse in a test
// https://storybook.js.org/docs/react/workflows/unit-testing
export const Default = Template.bind({});
Default.args = {
  loading: false,
  values: [...Array(6)].map(() => ''),
};

const UncontrolledTemplate: Story<UncontrolledDigitCodeInputProps> = (args) => (
  <UncontrolledDigitCodeInput {...args} />
);
export const Uncontrolled = UncontrolledTemplate.bind({});
Uncontrolled.args = {
  loading: false,
  length: 6,
  errorMessage: 'code is incorrect',
  onComplete: (code) => alert('Code is ' + code),
};
