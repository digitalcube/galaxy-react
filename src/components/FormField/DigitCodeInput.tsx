import React, {
  FC,
  useState,
  useEffect,
  useRef,
  PropsWithChildren,
  ReactNode,
} from 'react';

/**
 * @see https://codesandbox.io/s/verification-code-input-zjxvo?file=/src/index.js:267-488
 */
export type DigitCodeInputProps = PropsWithChildren<{
  loading: boolean;
  onChange: (values: string[]) => void;
  values: string[];
  errorMessage?: ReactNode;
}>;
export const DigitCodeInput: FC<DigitCodeInputProps> = ({
  loading,
  onChange,
  children,
  values,
  errorMessage,
}) => {
  const inputs = useRef<(HTMLInputElement | null)[]>([]);
  const length = values.length;

  const processInput = (
    e: React.ChangeEvent<HTMLInputElement>,
    slot: number
  ) => {
    const num = e.target.value;
    if (/[^0-9]/.test(num)) return;
    const newCode = [...values];
    newCode[slot] = num;
    onChange(newCode);
    if (slot !== length - 1) {
      const target = inputs.current[slot + 1];
      if (target) target.focus();
    }
  };

  const onKeyUp = (e: React.KeyboardEvent<HTMLInputElement>, slot: number) => {
    if (e.keyCode === 8 && !values[slot] && slot !== 0) {
      const newCode = [...values];
      newCode[slot - 1] = '';
      onChange(newCode);
      const target = inputs.current[slot - 1];
      if (target) target.focus();
    }
  };

  const pasteInput = (e: React.ClipboardEvent<HTMLInputElement>) => {
    //クリップボードの文字列を取得して配列に入れる
    let numbers = Array.from(e.clipboardData.getData('text'));
    //桁をチェックする
    const valuesLetters = values.length;
    const numbersLetters = numbers.length;
    if (valuesLetters === numbersLetters) {
      //inputを全て取得
      const inputBoxes = document.getElementsByName('num');
      for (let i = 0; i < inputBoxes.length; i++) {
        let inputBox: HTMLInputElement = inputBoxes[i] as HTMLInputElement;
        inputBox.value = String(numbers[i]);
      }
      let newCode = [...values];
      newCode = numbers;
      onChange(newCode);
    } else {
      alert('桁が違います');
    }
  };

  return (
    <div className="form-row mx-0 verification-code">
      {values.map((num, idx) => {
        return (
          <div className="col-2 px-0" key={idx}>
            <input
              className="form-control px-0 text-center"
              type="text"
              inputMode="numeric"
              maxLength={1}
              value={num}
              autoFocus={!values[0].length && idx === 0}
              readOnly={loading}
              onPaste={(e) => pasteInput(e)}
              onChange={(e) => processInput(e, idx)}
              onKeyUp={(e) => onKeyUp(e, idx)}
              ref={(ref) => inputs.current.push(ref)}
            />
          </div>
        );
      })}
      {errorMessage ? (
        <div className="d-block invalid-feedback">{errorMessage}</div>
      ) : null}
      {children}
    </div>
  );
};

export type UncontrolledDigitCodeInputProps = Omit<
  DigitCodeInputProps,
  'onChange' | 'values'
> & {
  onComplete: (value: string) => void;
  length: number;
};
export const UncontrolledDigitCodeInput: FC<UncontrolledDigitCodeInputProps> = (
  props
) => {
  const { length, onComplete } = props;
  const [values, setValues] = useState([...Array(length)].map(() => ''));
  useEffect(() => {
    if (values.every((num) => num !== '')) {
      onComplete(values.join(''));
    }
  }, [values, onComplete]);
  return <DigitCodeInput {...props} onChange={setValues} values={values} />;
};
