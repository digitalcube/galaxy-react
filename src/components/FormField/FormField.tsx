import React, { FC, ReactNode, useCallback, useEffect, useState } from 'react';
import {
  FormGroup,
  InputGroup,
  Label,
  Input,
  InputGroupAddon,
  InputProps,
} from 'reactstrap';
import { FaEye, FaEyeSlash } from 'react-icons/fa';
import { InputType } from 'reactstrap/es/Input';

export type FormFieldProps = {
  id: string;
  label: string;
  icon?: ReactNode;
  type: InputType;
  value: string;
  required?: boolean;
  readOnly?: boolean;
  disabled?: boolean;
  onChange: React.ChangeEventHandler<HTMLInputElement>;
  errorMessage?: string;
  inputOptions?: Omit<
    InputProps,
    'onChange' | 'type' | 'value' | 'id' | 'required' | 'readOnly' | 'disabled'
  >;
};

export const FormField: FC<FormFieldProps> = ({
  id,
  label,
  value,
  type,
  onChange,
  inputOptions,
  icon,
  required,
  readOnly,
  disabled,
  errorMessage,
}) => {
  const [inputType, setInputType] = useState(type);

  const toggleInputPasswordType = useCallback(() => {
    setInputType(inputType === 'password' ? 'text' : 'password');
  }, [inputType, setInputType]);

  useEffect(() => {
    setInputType(type);
  }, [type, setInputType]);
  const inputClassName = [
    inputOptions?.className,
    !!errorMessage ? 'is-invalid' : '',
  ]
    .filter(Boolean)
    .join(' ');
  const inputProps: InputProps = {
    type: inputType,
    value,
    id,
    className: inputClassName,
    placeholder: label,
    onChange,
    ...inputOptions,
  };
  if (required) inputProps.required = required;
  if (disabled) inputProps.disabled = disabled;
  if (readOnly) inputProps.readOnly = readOnly;
  return (
    <FormGroup className="mb-4">
      <Label className="font-weight-bold" htmlFor={id}>
        {label}
      </Label>
      <InputGroup className="reverse-icon">
        <Input {...inputProps} />
        {errorMessage ? (
          <div className="invalid-feedback">{errorMessage}</div>
        ) : null}
        {icon ? (
          <InputGroupAddon addonType="append">
            <i aria-hidden="true">{icon}</i>
          </InputGroupAddon>
        ) : null}
        {type === 'password' ? (
          <div
            className={`input-group-eye ${
              inputType === 'password' ? 'password-lock' : null
            }`}
            onClick={toggleInputPasswordType}
          >
            <i className="lock" aria-hidden="true">
              <FaEyeSlash className="d-block" />
            </i>
            <i className="unlock" aria-hidden="true">
              <FaEye className="d-block" />
            </i>
          </div>
        ) : null}
      </InputGroup>
    </FormGroup>
  );
};
