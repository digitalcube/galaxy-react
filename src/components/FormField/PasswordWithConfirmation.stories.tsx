import React from 'react';
import { Meta, Story } from '@storybook/react';
import { PasswordWithConfirmation, PasswordWithConfirmationProps } from './PasswordWithConfirmation';

const meta: Meta = {
  title: 'Components/Form/PasswordWithConfirmation',
  component: PasswordWithConfirmation,
  parameters: {
    controls: { expanded: true },
  },
};

export default meta;

const Template: Story<PasswordWithConfirmationProps> = (args) => (
  <>
    <PasswordWithConfirmation {...args} />
  </>
);

// By passing using the Args format for exported stories, you can control the props for a component for reuse in a test
// https://storybook.js.org/docs/react/workflows/unit-testing
export const Default = Template.bind({});
Default.args = {
    label: 'Choose a Password',
    id: 'password',
    value: '',
    confirmValue: '',
    onChange: () => undefined,
    onChangeConfirm: () => undefined,
    placeholder: 'Password',
    confirmPlaceholder: 'Re enter Password'
};
