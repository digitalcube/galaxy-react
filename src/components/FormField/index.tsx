export * from './FormField';
export * from './PasswordWithConfirmation';
export * from './Checkbox'