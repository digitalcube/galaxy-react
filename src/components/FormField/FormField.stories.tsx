import React from 'react';
import { Meta, Story } from '@storybook/react';
import { FormField, FormFieldProps } from './FormField';
import { FaUser } from 'react-icons/fa';

const meta: Meta = {
  title: 'Components/Form/FormField',
  component: FormField,
  parameters: {
    controls: { expanded: true },
  },
};

export default meta;

const Template: Story<FormFieldProps> = (args) => (
  <>
    <FormField {...args} />
  </>
);

// By passing using the Args format for exported stories, you can control the props for a component for reuse in a test
// https://storybook.js.org/docs/react/workflows/unit-testing
export const Default = Template.bind({});
Default.args = {
    icon: <FaUser className="d-block" />,
    id: 'password',
    label: 'Password',
    type: "password",
    onChange: () => undefined,
};
