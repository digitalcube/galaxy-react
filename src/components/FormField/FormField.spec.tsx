import { render } from '@testing-library/react';
import React from 'react';
import { Default as FormField } from './FormField.stories';

describe('FormField', () => {
  it('to match snapshot', () => {
    const { asFragment } = render(<FormField {...(FormField.args as any)} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('to match snapshot (with error massage)', () => {
    const { asFragment } = render(
        <FormField
            {...(FormField.args as any)}
            errorMessage="We donâ€™t recognize that email"
        />);
    expect(asFragment()).toMatchSnapshot();
  });
});
