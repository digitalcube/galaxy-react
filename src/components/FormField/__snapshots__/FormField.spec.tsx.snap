// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField to match snapshot (with error massage) 1`] = `
<DocumentFragment>
  <div
    class="mb-4 form-group"
  >
    <label
      class="font-weight-bold"
      for="password"
    >
      Password
    </label>
    <div
      class="reverse-icon input-group"
    >
      <input
        class="is-invalid form-control"
        id="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <div
        class="invalid-feedback"
      >
        We donâ€™t recognize that email
      </div>
      <div
        class="input-group-append"
      >
        <i
          aria-hidden="true"
        >
          <svg
            class="d-block"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
            />
          </svg>
        </i>
      </div>
      <div
        class="input-group-eye password-lock"
      >
        <i
          aria-hidden="true"
          class="lock"
        >
          <svg
            class="d-block"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 640 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z"
            />
          </svg>
        </i>
        <i
          aria-hidden="true"
          class="unlock"
        >
          <svg
            class="d-block"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
            />
          </svg>
        </i>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FormField to match snapshot 1`] = `
<DocumentFragment>
  <div
    class="mb-4 form-group"
  >
    <label
      class="font-weight-bold"
      for="password"
    >
      Password
    </label>
    <div
      class="reverse-icon input-group"
    >
      <input
        class="form-control"
        id="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <div
        class="input-group-append"
      >
        <i
          aria-hidden="true"
        >
          <svg
            class="d-block"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
            />
          </svg>
        </i>
      </div>
      <div
        class="input-group-eye password-lock"
      >
        <i
          aria-hidden="true"
          class="lock"
        >
          <svg
            class="d-block"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 640 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z"
            />
          </svg>
        </i>
        <i
          aria-hidden="true"
          class="unlock"
        >
          <svg
            class="d-block"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
            />
          </svg>
        </i>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
